### PR TITLE
Fix cast exception when running peon tasks(#16271)

### LINKED ISSUE #16271: MiddleManager errors while executing the peon task
MiddleManager errors while executing the peon task

### Affected Version

28.0.1

### Description

Please include as much detailed information about the problem as possible.
- Cluster size 
  3-node cluster
- Configurations in use 
 peon task context :
```
{"druid.indexer.fork.property.druid.processing.buffer.sizeBytes":33554432,"druid.indexer.fork.property.druid.processing.numThreads":3,"druid.indexer.runner.javaOpts":"-server -Xms256m -Xmx768m -XX:MaxDirectMemorySize=1g"}
```
- Steps to reproduce the problem
 When executing the Poen task, set druid-indexer.fork.property.druid.processing.buffer.sizeBytes to an integer

- The error message or stack traces encountered. Providing more context, such as nearby log messages or even entire logs, can be helpful.
```
2024-04-11 20:20:19,572 INFO  [forking-task-runner-35][][org.apache.druid.indexing.overlord.ForkingTaskRunner] Exception caught during execution
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
        at org.apache.druid.indexing.overlord.ForkingTaskRunner$1.call(ForkingTaskRunner.java:324) ~[druid-indexing-service-28.0.1-htrunk7.jar:?]
        at org.apache.druid.indexing.overlord.ForkingTaskRunner$1.call(ForkingTaskRunner.java:176) ~[druid-indexing-service-28.0.1-htrunk7.jar:?]
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131) ~[guava-31.1-jre.jar:?]
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74) ~[guava-31.1-jre.jar:?]
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82) ~[guava-31.1-jre.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_402]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_402]
        at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_402]
```
- Any debugging that you have already done
When the return value of task.getContextValue(propName) is an Integer instead of an int, "public CommandListBuilder addSystemProperty(String property, String value)" will be called and an error will occur: java. lang. ClassCastException: Java. lang. Integer cannot be cast to Java. lang. String.
I think it is necessary to change "public CommandListBuilder addSystemProperty(String property, int value)" to "public CommandListBuilder addSystemProperty(String property, Object value)" , and I will try to fix this bug soon
```
public class ForkingTaskRunner
                        final Map<String, Object> context = task.getContext();
                        if (context != null) {
                          for (String propName : context.keySet()) {
                            if (propName.startsWith(CHILD_PROPERTY_PREFIX)) {
                              command.addSystemProperty(
                                  propName.substring(CHILD_PROPERTY_PREFIX.length()),
                                  task.getContextValue(propName)
                              );
                            }
                          }
                        }
}
```
```
public static class CommandListBuilder
  {
    ArrayList<String> commandList = new ArrayList<>();

    public CommandListBuilder add(String arg)
    {
      commandList.add(arg);
      return this;
    }

    public CommandListBuilder addSystemProperty(String property, int value)
    {
      return addSystemProperty(property, String.valueOf(value));
    }

    public CommandListBuilder addSystemProperty(String property, long value)
    {
      return addSystemProperty(property, String.valueOf(value));
    }

    public CommandListBuilder addSystemProperty(String property, boolean value)
    {
      return addSystemProperty(property, String.valueOf(value));
    }

    public CommandListBuilder addSystemProperty(String property, String value)
    {
      return add(StringUtils.format("-D%s=%s", property, value));
    }

    public CommandListBuilder addAll(Iterable<String> args)
    {
      for (String arg : args) {
        add(arg);
      }
      return this;
    }

    public ArrayList<String> getCommandList()
    {
      return commandList;
    }

  }
```

### PR DESCRIPTION
**Pull request by soullkk**
Fixes #16271.

### Description
When calling method CommandListBuilder.addSystemProperty, if the parameter passed in is a generic type, then the actual method being called is CommandListBuilder addSystemProperty(String property, String value). If the actual type of the generic type is Integer, a cast exception will occur. So overloaded methods with parameters such as int, long, boolean, and String need to be removed, and used Object instead to solve runtime exceptions caused by generic type.

#### Release note
Fix the bug of cast exception when running peon task.

<hr>

##### Key changed/added classes in this PR
 * ForkingTaskRunner.CommandListBuilder

<hr>

This PR has:

- [X] been self-reviewed.
   - [X] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [X] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java [hunk 1]
```diff
@@ -909,24 +909,9 @@ public CommandListBuilder add(String arg)
       return this;
     }
 
-    public CommandListBuilder addSystemProperty(String property, int value)
+    public CommandListBuilder addSystemProperty(String property, Object value)
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:912]: Using an `Object` as the value of a system property is not appropriate. It must always be a primitive or a String. Hence, the different `addSystemProperty()` methods.

As a workaround to the original problem, you can simply pass the integer value quoted as a String. (https://github.com/apache/druid/issues/16271#issuecomment-2058155166)

Comment (soullkk) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:912]:
```
```
How about calling the toString() method to make the public CommandListBuilder addSystemProperty (String property, String value) called

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java [hunk 2]
```diff
@@ -316,7 +317,7 @@ public TaskStatus call()
                             if (propName.startsWith(CHILD_PROPERTY_PREFIX)) {
                               command.addSystemProperty(
                                   propName.substring(CHILD_PROPERTY_PREFIX.length()),
-                                  task.getContextValue(propName)
+                                  Preconditions.checkNotNull(task.getContextValue(propName), propName).toString()
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:320]: I don't think we should throw an exception. Maybe a property was set to `null` on purpose. We should just use the value of null.

Maybe something like this:
```java
if (propName.startsWith(CHILD_PROPERTY_PREFIX)) {
    final String contextValue = task.getContextValue(propName);
    command.addSystemProperty(
        propName.substring(CHILD_PROPERTY_PREFIX.length()),
        contextValue == null ? null : contextValue.toString()
    );
}    
```

**Review by kfaraz — APPROVED** [review]

**Comment by kfaraz** [comment]
@soullkk , the build is failing due to low coverage. Do you think you could add a test in `ForkingTaskRunnerTest` or update an existing test to cover these updated lines?

**Comment by soullkk** [comment]
> @soullkk , the build is failing due to low coverage. Do you think you could add a test in `ForkingTaskRunnerTest` or update an existing test to cover these updated lines?

ok， no problem.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java [hunk 3]
```diff
@@ -314,9 +314,10 @@ public TaskStatus call()
                         if (context != null) {
                           for (String propName : context.keySet()) {
                             if (propName.startsWith(CHILD_PROPERTY_PREFIX)) {
+                              Object contextValue = task.getContextValue(propName);
                               command.addSystemProperty(
                                   propName.substring(CHILD_PROPERTY_PREFIX.length()),
-                                  task.getContextValue(propName)
+                                  contextValue == null ? null : contextValue.toString()
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:320]: if the `contextValue` is null, does it make sense to call the `addSystemProperty` method?

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/ForkingTaskRunner.java:320]: Yeah, that's true. We can just skip the `addSystemProperty` if the value is null.

**Comment by maytasm** [comment]
> > @soullkk , the build is failing due to low coverage. Do you think you could add a test in `ForkingTaskRunnerTest` or update an existing test to cover these updated lines?
> 
> ok， no problem.

@soullkk A modified test like https://github.com/apache/druid/pull/16612/files#diff-399767f93acb71e3675e6ab1d6d62bd10e01d18d0405bc64e73a9ff993f9aadbR395 should work too. (just a FYI)

**Comment by soullkk** [comment]
> > > @soullkk , the build is failing due to low coverage. Do you think you could add a test in `ForkingTaskRunnerTest` or update an existing test to cover these updated lines?
> > 
> > 
> > ok， no problem.
> 
> @soullkk A modified test like https://github.com/apache/druid/pull/16612/files#diff-399767f93acb71e3675e6ab1d6d62bd10e01d18d0405bc64e73a9ff993f9aadbR395 should work too. (just a FYI)

ok, i will try it

**Comment by maytasm** [comment]
@soullkk We end up merging https://github.com/apache/druid/pull/16612 which reuses similar approach to this PR.

**Pull request closed without merging**